Feat/ava difficulty

Priyageorge123 opened the pull request:
Calculated all the difficulty metrics and the output json file is also attached.

Priyageorge123 committed:
Difficulty metrics and output json

Priyageorge123 committed:
Add files via upload

Priyageorge123 committed:
Deleted outdated json file

Priyageorge123 committed:
Updated json file

Priyageorge123 committed:
Enhance compute_difficulty function and output handling

Updated docstring for compute_difficulty function and added functionality to save output metrics to a JSON file.

adriandavidauer commented on examples/ava_tools/difficulty_from_csv.py:
```diff
@@ -0,0 +1,128 @@
+    areaA = max(0.0, ax2-ax1) * max(0.0, ay2-ay1)
+    areaB = max(0.0, bx2-bx1) * max(0.0, by2-by1)
+    union = areaA + areaB - inter + 1e-8
+    return inter / union
```
You can use compute_iou or compue_ious defined in paz/backend/boxes.py

adriandavidauer commented on examples/ava_tools/difficulty_from_csv.py:
```diff
@@ -0,0 +1,128 @@
+    union = areaA + areaB - inter + 1e-8
+    return inter / union
+
+def load_ava_as(csv_path):
```
please add docstrings for functions

adriandavidauer commented on examples/ava_tools/difficulty_from_csv.py:
```diff
@@ -0,0 +1,128 @@
+      - dynamics: average motion between frames (1 - IoU)
+      - occlusion: average overlap between faces
+      - audibility_sna_share: share of non-audible speaking labels
+    """
```
Please add paramaters and return to docstring using the style paz uses throught implementation

adriandavidauer commented on examples/ava_tools/difficulty_from_csv.py:
```diff
@@ -0,0 +1,128 @@
+    # audibility: share of SPEAKING_NOT_AUDIBLE among speaking frames
+    speak = [r for r in rows if r['label'].startswith('SPEAKING')]
+    sna = [r for r in speak if 'NOT_AUDIBLE' in r['label']]
+    audibility_sna_share = (len(sna) / len(speak)) if speak else 0.0
```
a percentage value would be easily computable with `len(sna) / (len(sna) + len(speak))`. I think this is better compareable between videos.

adriandavidauer commented on examples/ava_tools/difficulty_from_csv.py:
```diff
@@ -0,0 +1,128 @@
+        for i in range(1, len(seq)):
+            iou_val = iou(seq[i-1]['box'], seq[i]['box'])
+            dynamics_terms.append(1.0 - iou_val)
+    dynamics = mean(dynamics_terms) if dynamics_terms else 0.0
```
here maybe mean and variance are better. Long sequences with only a short period of fast movement will get a very low mean but variance would be greater than for a sequence where only little movement happens.

adriandavidauer commented on examples/ava_tools/difficulty_from_csv.py:
```diff
@@ -0,0 +1,128 @@
+        for i in range(n):
+            for j in range(i+1, n):
+                s += iou(boxes[i], boxes[j])
+                c += 1
```
I think this can be done with `compute_ious` from paz 👍

adriandavidauer commented on examples/ava_tools/difficulty_from_csv.py:
```diff
@@ -0,0 +1,128 @@
+                s += iou(boxes[i], boxes[j])
+                c += 1
+        if c:
+            occlusion_terms.append(s / c)
```
I do not get what c is doing - it is the number of combinations between detections if I'm correct. Why don't you just add all IOUs to the occlusion_terms? So you basically average for each frame and then later you average over all averages. Correct me if I'm wrong but I think that can be easier.

adriandavidauer commented on examples/ava_tools/difficulty_from_csv.py:
```diff
@@ -0,0 +1,128 @@
+
+if __name__ == "__main__":
+    import sys, os, json, glob
+    path = sys.argv[1]  
```
please use argparse

adriandavidauer commented on examples/ava_tools/difficulty_from_csv.py:
```diff
@@ -0,0 +1,128 @@
+        for csv_path in sorted(glob.glob(os.path.join(path, "*.csv"))):
+            process_csv(csv_path)
+
+    # Save JSON next to this script
```
next to the script can be the default but please use argparse to make this more flexible

adriandavidauer reviewed (changes requested):
Looks good so far some small adjustements and we still need to think about normalization for the dificulty values so that they are somewhat compareable.

Priyageorge123 committed:
Add docstrings to functions in difficulty_from_csv.py

Added docstrings to load_ava_as and compute_difficulty functions for better clarity on their arguments and return values.

Priyageorge123 committed:
Refactored difficulty computation and added augment parsing

Updated difficulty metrics computation and added argument parsing for CSV files.

Priyageorge123 committed:
Refactored argument parser for output file handling

adriandavidauer reviewed (approved):
Nicely done!
